mmv1/third_party/terraform/services/storage/resource_storage_folder.go.tmpl

```diff
@@ -0,0 +1,358 @@
+	}
+
+	log.Printf("[DEBUG] Finished creating Folder %q: %#v", d.Id(), res)
+
+	return resourceStorageFolderRead(d, meta)
+}
+
+func resourceStorageFolderRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*transport_tpg.Config)
+	userAgent, err := tpgresource.GenerateUserAgentString(d, config.UserAgent)
+	if err != nil {
+		return err
+	}
+	bucket := d.Get("bucket").(string)
+	folder := d.Get("name").(string)
+	var res *storage.Folder
+	err = transport_tpg.Retry(transport_tpg.RetryOptions{
+		RetryFunc: func() (operr error) {
+			var retryErr error
+			res, retryErr = config.NewStorageClient(userAgent).Folders.Get(bucket, folder).Do()
+			return retryErr
+		},
+		Timeout:              d.Timeout(schema.TimeoutRead),
+		ErrorRetryPredicates: []transport_tpg.RetryErrorPredicateFunc{transport_tpg.IsNotFoundRetryableError("folder read")},
+	})
+	if err != nil {
+		return transport_tpg.HandleNotFoundError(err, d, fmt.Sprintf("StorageFolder %q", d.Id()))
+	}
+
+	if err := d.Set("create_time", flattenStorageFolderCreateTime(res.CreateTime, d, config)); err != nil {
+		return fmt.Errorf("Error reading Folder: %s", err)
+	}
+	if err := d.Set("update_time", flattenStorageFolderUpdateTime(res.UpdateTime, d, config)); err != nil {
+		return fmt.Errorf("Error reading Folder: %s", err)
+	}
+	if err := d.Set("metageneration", flattenStorageFolderMetageneration(res.Metageneration, d, config)); err != nil {
+		return fmt.Errorf("Error reading Folder: %s", err)
+	}
+	if err := d.Set("bucket", flattenStorageFolderBucket(res.Bucket, d, config)); err != nil {
+		return fmt.Errorf("Error reading Folder: %s", err)
+	}
+	if err := d.Set("name", flattenStorageFolderName(res.Name, d, config)); err != nil {
+		return fmt.Errorf("Error reading Folder: %s", err)
+	}
+	if err := d.Set("self_link", tpgresource.ConvertSelfLinkToV1(res.SelfLink)); err != nil {
+		return fmt.Errorf("Error reading Folder: %s", err)
+	}
+
+	return nil
+}
+
+func resourceStorageFolderUpdate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*transport_tpg.Config)
+	_ = config
+
+	// we can only get here if force_destroy was updated
+	if d.Get("force_destroy") != nil {
+		if err := d.Set("force_destroy", d.Get("force_destroy")); err != nil {
+			return fmt.Errorf("Error updating force_destroy: %s", err)
+		}
+	}
+
+	// all other fields are immutable, don't do anything else
+	return nil
+}
+
+func resourceStorageFolderDelete(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*transport_tpg.Config)
+	userAgent, err := tpgresource.GenerateUserAgentString(d, config.UserAgent)
+	if err != nil {
+		return err
+	}
+
+	bucket := d.Get("bucket").(string)
+	name := d.Get("name").(string)
+
+	folderName := strings.Split(name, "/")
+	var listError, deleteObjectError error
+	for deleteObjectError == nil {
+		res, err := config.NewStorageClient(userAgent).Objects.List(bucket).Prefix(folderName[0] + "/").Do()
```

Comment:
Why to list objects of the top level folder when we are trying to delete nested folder?

Reply:
to delete folders first we need to delete all objects as non empty folders cannot be deleted, getting all the objects with top level folder prefix and delete all objects and in next step delete the sub folders and finally delete the main folder, this will clear up all the objects, subfolders and folders

Reply:
But the scenario that I am referring is, If the folder resource is `a/b/c/d` then we try to delete it with `force_destroy` = true then as per this logic, it will try to delete everything inside folder `a/` which is not expected. we should only delete everything that is nested in `a/b/c/d` path not anything that is in upper level of path or in different path tree like `a/b/c` or `a/b/c/e/f`.

Reply:
updated the logic